File: week41_250317/서진경/BOJ_250318_16938.java
package beakjoon.gold;

import java.io.*;
import java.util.*;

public class boj_16938_캠프준비 {
    static BufferedReader br;
    static StringTokenizer st;
    static int[] arr;
    static int N, L, R, X, res;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        L = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());
        X = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        arr = new int[N];
        for(int n=0; n<N; n++) {
            arr[n] = Integer.parseInt(st.nextToken());
        }

        res = 0;
        selectNumber(0, 0, Integer.MAX_VALUE, Integer.MIN_VALUE);
        System.out.println(res);
    }

    public static void selectNumber(int idx, int sum, int min, int max) {
        if(sum > R || idx >= N) return;

        if(Math.max(max, arr[idx])-Math.min(min, arr[idx]) >= X && sum+arr[idx] >= L && sum+arr[idx]<=R) res++;
        selectNumber(idx+1, sum+arr[idx], Math.min(min, arr[idx]), Math.max(max, arr[idx]));
        selectNumber(idx+1, sum, min, max);
    }
}

File: week41_250317/박준영/BOJ_250318_16938.py
import sys
input = sys.stdin.readline

N, L, R, X = map(int, input().split())
arr = list(map(int, input().split()))

count = 0

for mask in range(1, 1<<N):
    selected = [arr[i] for i in range(N) if mask & (1<<i)]
    
    if len(selected) < 2:
        continue
    
    total = sum(selected)
    min_value = min(selected)
    max_value = max(selected)
    
    if L <= total <= R and (max_value-min_value) >= X:
        count += 1

print(count)
    